src/python/pants/backend/java/bsp/rules.py

```diff
@@ -0,0 +1,221 @@
+    Target,
+    WrappedTarget,
+)
+from pants.engine.unions import UnionMembership, UnionRule
+from pants.jvm.bsp.spec import JvmBuildTarget
+from pants.jvm.compile import ClasspathEntryRequest, FallibleClasspathEntry
+from pants.jvm.resolve.key import CoursierResolveKey
+
+LANGUAGE_ID = "java"
+
+_logger = logging.getLogger(__name__)
+
+
+class JavaBSPLanguageSupport(BSPLanguageSupport):
+    language_id = LANGUAGE_ID
+    can_compile = True
+
+
+class JavaBSPBuildTargetsRequest(BSPBuildTargetsRequest):
+    pass
+
+
+@dataclass(frozen=True)
+class ResolveJavaBSPBuildTargetRequest:
+    target: Target
+
+
+@rule
+async def bsp_resolve_one_java_build_target(
+    request: ResolveJavaBSPBuildTargetRequest,
+    union_membership: UnionMembership,
+) -> BuildTarget:
+    dep_addrs = await Get(Addresses, DependenciesRequest(request.target[Dependencies]))
+    impls = union_membership.get(BSPCompileFieldSet)
+
+    reported_deps = []
+    for dep_addr in dep_addrs:
+        if dep_addr == request.target.address:
+            continue
+
+        wrapped_dep_tgt = await Get(WrappedTarget, Address, dep_addr)
+        dep_tgt = wrapped_dep_tgt.target
+        for impl in impls:
+            if impl.is_applicable(dep_tgt):
+                reported_deps.append(BuildTargetIdentifier.from_address(dep_tgt.address))
+                break
+
+    return BuildTarget(
+        id=BuildTargetIdentifier.from_address(request.target.address),
+        display_name=str(request.target.address),
+        base_directory=None,
+        tags=(),
+        capabilities=BuildTargetCapabilities(
+            can_compile=True,
+        ),
+        language_ids=(LANGUAGE_ID,),
+        dependencies=tuple(reported_deps),
+        data_kind="jvm",
+        data=JvmBuildTarget(),
+    )
+
+
+@rule
+async def bsp_resolve_all_java_build_targets(
+    _: JavaBSPBuildTargetsRequest,
+    all_java_targets: AllJavaTargets,
+    bsp_context: BSPContext,
+) -> BSPBuildTargets:
+    if LANGUAGE_ID not in bsp_context.client_params.capabilities.language_ids:
+        return BSPBuildTargets()
+    build_targets = await MultiGet(
+        Get(BuildTarget, ResolveJavaBSPBuildTargetRequest(tgt)) for tgt in all_java_targets
+    )
+    return BSPBuildTargets(targets=tuple(build_targets))
+
+
+# -----------------------------------------------------------------------------------------------
+# Javac Options Request
+# See https://build-server-protocol.github.io/docs/extensions/java.html#javac-options-request
+# -----------------------------------------------------------------------------------------------
+
+
+class JavacOptionsHandlerMapping(BSPHandlerMapping):
+    method_name = "buildTarget/javacOptions"
+    request_type = JavacOptionsParams
+    response_type = JavacOptionsResult
+
+
+@dataclass(frozen=True)
+class HandleJavacOptionsRequest:
+    bsp_target_id: BuildTargetIdentifier
+
+
+@dataclass(frozen=True)
+class HandleScalacOptionsResult:
+    item: JavacOptionsItem
```

Comment:
Needs rename.

It's not clear that it needs to be wrapped though?

Reply:
Probably not, but an optimization that can be applied once I have a MVP working.